### PR TITLE
add validation of required and default values

### PR DESCRIPTION
**Pull request by campoy**

### CODE REVIEW

**Comment by campoy** [comment]
Please review @crhym3

**Comment by crhym3** [comment]
Python version never implemented it so neither did I. I think it was supposed to be on a lower lever, API infra, but not entirely sure.
@campoy thanks for doing this.

**Comment by campoy** [comment]
PTAL @crhym3 @okdave

File: endpoints/apiconfig.go [hunk 1]
```diff
@@ -744,3 +744,46 @@ func parseValue(s string, k reflect.Kind) (interface{}, error) {
...
+	}
+	return nil
+}
+
+func validateField(v reflect.Value, t reflect.StructField) error {
+	tag, err := parseTag(t.Tag)
+	if err != nil {
+		return fmt.Errorf("parse tag: %v", err)
+	}
+	zero := reflect.Zero(v.Type()).Interface()
```
Comment (okdave) [inline endpoints/apiconfig.go]: Now that you've inlined all this, how about ditching earlier on non-zero?

if zero := reflect.Zero(v.Type()).Interface(); zero != v.Interface() {
  return nil
}

if tag.required { // etc. }}

if tag.default == "" {
  return nil
}

etc.

**Comment by okdave** [comment]
Looking good!

File: endpoints/apiconfig.go (hunk 1 above)
Comment (campoy) [inline endpoints/apiconfig.go]: done, thanks!

**Comment by kylefinley** [comment]
@campoy Nice work! I'm happy to see this implemented.

I believe this commit maybe breaking some existing behavior.

Prior to this commit it was possible for a `POST` or `PUT` request to contain a slice, e.g.

```go
type TestReq struct {
	Strings []string `json:"strings"`
}

func (s *Service) Test(c context.Context, req *TestReq) error {
	return nil
}
```

But now if I attempt that I'm seeing a panic:

```bash
2015/07/29 19:30:21 http: panic serving 127.0.0.1:62097: runtime error: comparing uncomparable type []string
goroutine 6 [running]:
net/http.func·011()
	/private/var/folders/00/0v42r000h01000cxqpysvccm003chb/T/appengine/go_appengine/goroot/src/net/http/server.go:1130 +0xbb
github.com/GoogleCloudPlatform/go-endpoints/endpoints.validateField(0x3697c0, 0xc2080ee840, 0xd7, 0x4dcc60, 0x7, 0x0, 0x0, 0x85aa58, 0x3697c0, 0x5054b0, ...)
	/dev/go/src/github.com/GoogleCloudPlatform/go-endpoints/endpoints/apiconfig.go:773 +0x27e
github.com/GoogleCloudPlatform/go-endpoints/endpoints.validateRequest(0x3633a0, 0xc2080ee840, 0x0, 0x0)
	/dev/go/src/github.com/GoogleCloudPlatform/go-endpoints/endpoints/apiconfig.go:760 +0x438
github.com/GoogleCloudPlatform/go-endpoints/endpoints.(*Server).ServeHTTP(0xc20807ffc0, 0x85c528, 0xc208044d20, 0xc208033520)
	/dev/go/src/github.com/GoogleCloudPlatform/go-endpoints/endpoints/server.go:151 +0x937
net/http.(*ServeMux).ServeHTTP(0xc20803a690, 0x85c528, 0xc208044d20, 0xc208033520)
	/private/var/folders/00/0v42r000h01000cxqpysvccm003chb/T/appengine/go_appengine/goroot/src/net/http/server.go:1541 +0x17d
appengine_internal.handleFilteredHTTP(0x85c528, 0xc208044d20, 0xc208033520)
	/private/var/folders/00/0v42r000h01000cxqpysvccm003chb/T/appengine/go_appengine/goroot/src/appengine_internal/api_dev.go:98 +0x413
net/http.HandlerFunc.ServeHTTP(0x5dc048, 0x85c528, 0xc208044d20, 0xc208033520)
	/private/var/folders/00/0v42r000h01000cxqpysvccm003chb/T/appengine/go_appengine/goroot/src/net/http/server.go:1265 +0x41
net/http.serverHandler.ServeHTTP(0xc2080421e0, 0x85c528, 0xc208044d20, 0xc208033520)
	/private/var/folders/00/0v42r000h01000cxqpysvccm003chb/T/appengine/go_appengine/goroot/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc208044c80)
	/private/var/folders/00/0v42r000h01000cxqpysvccm003chb/T/appengine/go_appengine/goroot/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
	/private/var/folders/00/0v42r000h01000cxqpysvccm003chb/T/appengine/go_appengine/goroot/src/net/http/server.go:1751 +0x35e
INFO     2015-07-29 19:30:21,200 module.py:809] default: "POST /_ah/spi/Service.Test HTTP/1.1" 500 2399
INFO     2015-07-29 19:30:21,201 module.py:809] default: "PUT /_ah/api/test/v1/test?alt=json HTTP/1.1" 500 2490
INFO     2015-07-29 19:33:02,482 module.py:442] [default] Detected file changes:
```

Any ideas on what might be causing this. Thanks!

**Comment by campoy** [comment]
#96 filed and PR sent to fix it

Thanks!

**Pull request merged**